ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -312,9 +312,39 @@ let pool_migrate ~__context ~vm ~host ~options =
   let queue_name = queue_of_vm ~__context ~self:vm in
   let module XenopsAPI = (val make_client queue_name : XENOPS) in
   let session_id = Ref.string_of (Context.get_session_id __context) in
-  let ip =
-    Http.Url.maybe_wrap_IPv6_literal (Db.Host.get_address ~__context ~self:host)
+  (* If `network` provided in `options`, try to get `xenops_url` on this network *)
+  let address =
+    match List.assoc_opt "network" options with
+    | None ->
+        Db.Host.get_address ~__context ~self:host
+    | Some network_ref -> (
+        let network = Ref.of_string network_ref in
+        match
+          Xapi_network_attach_helpers.get_local_pifs ~__context ~network ~host
+        with
+        | [] ->
+            raise
+              Api_errors.(
+                Server_error
+                  (pif_not_present, [Ref.string_of host; Ref.string_of network]))
+        | pifs -> (
+          try
+            let pif =
+              List.find
+                (fun pif ->
+                  Xapi_pif_helpers.get_primary_address ~__context ~pif <> None)
+                pifs
+            in
+            Option.get (Xapi_pif_helpers.get_primary_address ~__context ~pif)
+          with Not_found ->
+            failwith
+              (Printf.sprintf
+                 "No PIF with IP configuration for host: [%s] on network: [%s]"
+                 (Ref.string_of host) (Ref.string_of network))
+        )
```

Comment:
By definition, there will only be zero or one local PIF for a host on any given network.

```suggestion
        | pif :: _ -> (
          match Xapi_pif_helpers.get_primary_address ~__context ~pif with
          | Some ip -> ip
          | None ->
            raise
              Api_errors.(Server_error (interface_has_no_ip, [Ref.string_of pif]))
        )
```